File: deob/src/main/java/dev/openrs2/deob/analysis/IntInterpreter.java
package dev.openrs2.deob.analysis;

import java.util.List;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Sets;
import dev.openrs2.asm.InsnNodeUtils;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.IincInsnNode;
import org.objectweb.asm.tree.analysis.AnalyzerException;
import org.objectweb.asm.tree.analysis.BasicInterpreter;
import org.objectweb.asm.tree.analysis.BasicValue;
import org.objectweb.asm.tree.analysis.Interpreter;

public final class IntInterpreter extends Interpreter<IntValue> {
	// TODO(gpe): this is fairly arbitrary and will need tweaking
	private static final int MAX_TRACKED_VALUES = 8;

	private final Interpreter<BasicValue> basicInterpreter = new BasicInterpreter();
	private final ImmutableSet<Integer>[] parameters;

	public IntInterpreter(ImmutableSet<Integer>[] parameters) {
		super(Opcodes.ASM7);
		this.parameters = parameters;
	}

	@Override
	public IntValue newValue(Type type) {
		var basicValue = basicInterpreter.newValue(type);
		if (basicValue == null) {
			return null;
		}

		return IntValue.newUnknown(basicValue);
	}

	@Override
	public IntValue newParameterValue(boolean isInstanceMethod, int local, Type type) {
		var basicValue = basicInterpreter.newParameterValue(isInstanceMethod, local, type);
		if (basicValue == null) {
			return null;
		}

		if (parameters != null) {
			int parameterIndex;
			if (isInstanceMethod) {
				if (local == 0) {
					return IntValue.newUnknown(basicValue);
				}
				parameterIndex = local - 1;
			} else {
				parameterIndex = local;
			}

			var parameter = parameters[parameterIndex];
			if (parameter != null) {
				return IntValue.newConstant(basicValue, parameter);
			}
		}

		return IntValue.newUnknown(basicValue);
	}

	@Override
	public IntValue newOperation(AbstractInsnNode insn) throws AnalyzerException {
		var basicValue = basicInterpreter.newOperation(insn);
		if (basicValue == null) {
			return null;
		}

		if (InsnNodeUtils.isIntConstant(insn)) {
			return IntValue.newConstant(basicValue, InsnNodeUtils.getIntConstant(insn));
		}

		return IntValue.newUnknown(basicValue);
	}

	@Override
	public IntValue copyOperation(AbstractInsnNode insn, IntValue value) throws AnalyzerException {
		return value;
	}

	@Override
	public IntValue unaryOperation(AbstractInsnNode insn, IntValue value) throws AnalyzerException {
		var basicValue = basicInterpreter.unaryOperation(insn, value.getBasicValue());
		if (basicValue == null) {
			return null;
		}

		if (value.isUnknown()) {
			return IntValue.newUnknown(basicValue);
		}

		var set = ImmutableSet.<Integer>builder();

		for (var v : value.getIntValues()) {
			switch (insn.getOpcode()) {
			case Opcodes.INEG:
				set.add(-v);
				break;
			case Opcodes.IINC:
				var iinc = (IincInsnNode) insn;
				set.add(v + iinc.incr);
				break;
			case Opcodes.I2B:
				set.add((int) (byte) (int) v);
				break;
			case Opcodes.I2C:
				set.add((int) (char) (int) v);
				break;
			case Opcodes.I2S:
				set.add((int) (short) (int) v);
				break;
			default:
				return IntValue.newUnknown(basicValue);
			}
		}

		return IntValue.newConstant(basicValue, set.build());
	}

	@Override
	public IntValue binaryOperation(AbstractInsnNode insn, IntValue value1, IntValue value2) throws AnalyzerException {
		var basicValue = basicInterpreter.binaryOperation(insn, value1.getBasicValue(), value2.getBasicValue());
		if (basicValue == null) {
			return null;
		}

		if (value1.isUnknown() || value2.isUnknown()) {
			return IntValue.newUnknown(basicValue);
		}

		var set = ImmutableSet.<Integer>builder();

		for (var v1 : value1.getIntValues()) {
			for (var v2 : value2.getIntValues()) {
				switch (insn.getOpcode()) {
				case Opcodes.IADD:
					set.add(v1 + v2);
					break;
				case Opcodes.ISUB:
					set.add(v1 - v2);
					break;
				case Opcodes.IMUL:
					set.add(v1 * v2);
					break;
				case Opcodes.IDIV:
					if (v2 == 0) {
						return IntValue.newUnknown(basicValue);
					}
					set.add(v1 / v2);
					break;
				case Opcodes.IREM:
					if (v2 == 0) {
						return IntValue.newUnknown(basicValue);
					}
					set.add(v1 % v2);
					break;
				case Opcodes.ISHL:
					set.add(v1 << v2);
					break;
				case Opcodes.ISHR:
					set.add(v1 >> v2);
					break;
				case Opcodes.IUSHR:
					set.add(v1 >>> v2);
					break;
				case Opcodes.IAND:
					set.add(v1 & v2);
					break;
				case Opcodes.IOR:
					set.add(v1 | v2);
					break;
				case Opcodes.IXOR:
					set.add(v1 ^ v2);
					break;
				default:
					return IntValue.newUnknown(basicValue);
				}
			}
		}

		return IntValue.newConstant(basicValue, set.build());
	}

	@Override
	public IntValue ternaryOperation(AbstractInsnNode insn, IntValue value1, IntValue value2, IntValue value3) throws AnalyzerException {
		var basicValue = basicInterpreter.ternaryOperation(insn, value1.getBasicValue(), value2.getBasicValue(), value3.getBasicValue());
		if (basicValue == null) {
			return null;
		}

		return IntValue.newUnknown(basicValue);
	}

	@Override
	public IntValue naryOperation(AbstractInsnNode insn, List<? extends IntValue> values) throws AnalyzerException {
		var args = values.stream()
			.map(IntValue::getBasicValue)
			.collect(ImmutableList.toImmutableList());
		var basicValue = basicInterpreter.naryOperation(insn, args);
		if (basicValue == null) {
			return null;
		}

		return IntValue.newUnknown(basicValue);
	}

	@Override
	public void returnOperation(AbstractInsnNode insn, IntValue value, IntValue expected) throws AnalyzerException {
		basicInterpreter.returnOperation(insn, value.getBasicValue(), expected.getBasicValue());
	}

	@Override
	public IntValue merge(IntValue value1, IntValue value2) {
		var basicValue = basicInterpreter.merge(value1.getBasicValue(), value2.getBasicValue());
		if (basicValue == null) {
			return null;
		}

		if (value1.isUnknown() || value2.isUnknown()) {
			return IntValue.newUnknown(basicValue);
		}

		var set = ImmutableSet.copyOf(Sets.union(value1.getIntValues(), value2.getIntValues()));
		if (set.size() > MAX_TRACKED_VALUES) {
			return IntValue.newUnknown(basicValue);
		}

		return IntValue.newConstant(basicValue, set);
	}
}

File: deob/src/main/java/dev/openrs2/deob/analysis/IntInterpreter.kt
package dev.openrs2.deob.analysis

import dev.openrs2.asm.intConstant
import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.IincInsnNode
import org.objectweb.asm.tree.analysis.BasicInterpreter
import org.objectweb.asm.tree.analysis.Interpreter

class IntInterpreter(private val args: Array<IntValueSet>) : Interpreter<IntValue>(Opcodes.ASM8) {
    private val basicInterpreter = BasicInterpreter()

    override fun newValue(type: Type?): IntValue? {
        val basicValue = basicInterpreter.newValue(type) ?: return null
        return IntValue(basicValue)
    }

    override fun newParameterValue(isInstanceMethod: Boolean, local: Int, type: Type): IntValue {
        val basicValue = basicInterpreter.newParameterValue(isInstanceMethod, local, type)

        val index = when {
            isInstanceMethod && local == 0 -> return IntValue(basicValue)
            isInstanceMethod -> local - 1
            else -> local
        }

        return IntValue(basicValue, args[index])
    }

    override fun newOperation(insn: AbstractInsnNode): IntValue {
        val basicValue = basicInterpreter.newOperation(insn)
        val v = insn.intConstant
        return if (v != null) {
            IntValue(basicValue, IntValueSet.singleton(v))
        } else {
            IntValue(basicValue)
        }
    }

    override fun copyOperation(insn: AbstractInsnNode, value: IntValue): IntValue {
        return value
    }

    override fun unaryOperation(insn: AbstractInsnNode, value: IntValue): IntValue? {
        val basicValue = basicInterpreter.unaryOperation(insn, value.basicValue) ?: return null

        if (value.set !is IntValueSet.Constant) {
            return IntValue(basicValue)
        }

        val set = mutableSetOf<Int>()
        for (v in value.set.values) {
            val result = when {
                insn.opcode == Opcodes.INEG -> -v
                insn is IincInsnNode -> v + insn.incr
                insn.opcode == Opcodes.I2B -> v.toByte().toInt()
                insn.opcode == Opcodes.I2C -> v.toChar().toInt()
                insn.opcode == Opcodes.I2S -> v.toShort().toInt()
                else -> return IntValue(basicValue)
            }
            set.add(result)
        }
        return IntValue(basicValue, IntValueSet.Constant(set))
    }

    override fun binaryOperation(insn: AbstractInsnNode, value1: IntValue, value2: IntValue): IntValue? {
        val basicValue = basicInterpreter.binaryOperation(insn, value1.basicValue, value2.basicValue) ?: return null

        if (value1.set !is IntValueSet.Constant || value2.set !is IntValueSet.Constant) {
            return IntValue(basicValue)
        }

        val set = mutableSetOf<Int>()
        for (v1 in value1.set.values) {
            for (v2 in value2.set.values) {
                val result = when (insn.opcode) {
                    Opcodes.IADD -> v1 + v2
                    Opcodes.ISUB -> v1 - v2
                    Opcodes.IMUL -> v1 * v2
                    Opcodes.IDIV -> {
                        if (v2 == 0) {
                            return IntValue(basicValue)
                        }
                        v1 / v2
                    }
                    Opcodes.IREM -> {
                        if (v2 == 0) {
                            return IntValue(basicValue)
                        }
                        v1 % v2
                    }
                    Opcodes.ISHL -> v1 shl v2
                    Opcodes.ISHR -> v1 shr v2
                    Opcodes.IUSHR -> v1 ushr v2
                    Opcodes.IAND -> v1 and v2
                    Opcodes.IOR -> v1 or v2
                    Opcodes.IXOR -> v1 xor v2
                    else -> return IntValue(basicValue)
                }
                set.add(result)
            }
        }
        return IntValue(basicValue, IntValueSet.Constant(set))
    }

    override fun ternaryOperation(
        insn: AbstractInsnNode,
        value1: IntValue,
        value2: IntValue,
        value3: IntValue
    ): IntValue? {
        val basicValue =
            basicInterpreter.ternaryOperation(insn, value1.basicValue, value2.basicValue, value3.basicValue)
                ?: return null
        return IntValue(basicValue)
    }

    override fun naryOperation(insn: AbstractInsnNode, values: List<IntValue>): IntValue? {
        val args = values.map(IntValue::basicValue)
        val basicValue = basicInterpreter.naryOperation(insn, args) ?: return null
        return IntValue(basicValue)
    }

    override fun returnOperation(insn: AbstractInsnNode, value: IntValue, expected: IntValue) {
        basicInterpreter.returnOperation(insn, value.basicValue, expected.basicValue)
    }

    override fun merge(value1: IntValue, value2: IntValue): IntValue {
        val basicValue = basicInterpreter.merge(value1.basicValue, value2.basicValue)

        if (value1 == value2) {
            return value1
        }

        if (value1.set !is IntValueSet.Constant || value2.set !is IntValueSet.Constant) {
            return IntValue(basicValue)
        }

        val set = value1.set union value2.set
        return if (set is IntValueSet.Constant && set.values.size > MAX_TRACKED_VALUES) {
            IntValue(basicValue)
        } else {
            IntValue(basicValue, set)
        }
    }

    companion object {
        private const val MAX_TRACKED_VALUES = 1
    }
}
